### ISSUE TITLE
"Do not evict" only for running pods

### ISSUE THREAD

**Issue by runningman84** [aws/karpenter#3537]
### Tell us about your request

We configure all of our helm hooks and other important pods with the do not evict flag. Some pods might not move to the running state because of having errors like CreateContainerError, CreateContainerConfigError or others.

Therefore we would like to have a better karpenter behaviour, only consider the do not evict flag if a given pod is in state creating or running...  

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

Right now a single pod can keep a single node running for unlimited time. After some time this is very costly... Imagine having dozens of nodes in a left behind state. 

### Are you currently working around this issue?

Manually kill these pods. Which does not work if you have dozens of teams and projects running on a given cluster...

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by jonathan-innis** [comment]
I think in general it makes sense if the pod is in an errored state and is clearly not running, that disruption like this should be allowed.

One thing I can think of is that this has the potential to race and put clusters in a half-state if they have jobs that are stateful:
1. Pod errors and enters a Failed state
2. `do-not-evict` is ignored and Karpenter begins to execute deprovisioning
3. Pod starts running
4. Node is deprovisioned which leaves the job half-done

Any thoughts to add to this @njtran?

**Comment by jonathan-innis** [comment]
> This race conditions would only occur if a given pod is the only reason why the cluster scaled up in the first place

Not neccesarily. Consolidation or expiration could occur on a node with `do-not-evict` pods and there would still be other workloads that would be scheduled on that node.

I think general guidance is that jobs should be able to be disrupted and non-stateful since there's any number of reasons why they could fail or get rescheduled.

**Comment by runningman84** [comment]
There won’t be other new workloads because once karpenter decides to remove it the node is set to no schedule anyway.

**Comment by njtran** [comment]
So I believe that we already check if a pod is terminal (succeeded or failed) when considering the do-not-evict pod. 

https://github.com/aws/karpenter-core/blob/main/pkg/controllers/deprovisioning/helpers.go#L328-L334
https://github.com/aws/karpenter-core/blob/main/pkg/utils/pod/scheduling.go#L50-L52

Is there something I'm missing here? are you not seeing this happening in your cluster? In the case of only considering running or creating pods, I'm not sure if we should ignore the `do-not-evict` in the case of Unknown or Terminating, as those map to many different scenarios as well. If I understand correctly, a pod being terminal (succeeded or failed) should be sufficient.

**Comment by njtran** [comment]
After poking around, I think I understand your request now. Looks like you want to bypass the annotation in the case that any of the containers have a problem that you deem a "terminal" state, resulting in the pod staying in the pending phase. If we wanted to try this, I think we'd have to check all the containers on the pod and allow a set of container statuses to bypass the annotation. 

Specifically checking the go struct used for this, it doesn't seem like there's a normal set of enum values for this: https://pkg.go.dev/k8s.io/api/core/v1#ContainerStateWaiting
So this means "Waiting" could happen for any number of reasons that may remediate itself, or not, tough to know. 

I'd be happy to support this though, if we can construct a specific set of container status reasons that we can deem to satisfy [@jonathan-innis's condition](https://github.com/aws/karpenter/issues/3537#issuecomment-1456663990).
> pod is in an errored state and is clearly not running

**Comment by runningman84** [comment]
Yes I would like to ignore pods which have containers wich will never recover from a terminal state but I would also like to simple ignore pods which are in these states:
* Failed
* Succeeded/Completed (this could already be implemented?)

Maybe for pods in a pending state we could ignore them after some "timeout"?

**Comment by jonathan-innis** [comment]
> The pod is in pending state and the container is in waiting state, here is a similar example

Is it possible to get this at the source and remediate the containers that are hanging in this state so that you don't get hung nodes? Is it possible to create some sort of alerting around how long a Pod sits in a Pending phase so that if one sits there for too long, you get an alarm and then can go to the dev team to ask them to remediate their image/container?

**Comment by runningman84** [comment]
We have alerting in place, but if you have a lot of teams using the cluster it is a pain to manually take care of these issues.
